File: barbershop/bs.go
// This is an implementation using Go channels of the solution described in the Little Book of Semaphores.
package main

import (
	"fmt"
	"sync"
	"time"
)

const maxCustomers = 4     // Maximum number of customers allowed in the shop.
const totalCustomers = 100 // Total number of customers that will arrive throughout the program.
const numTrials = 100

var waitingTimes [totalCustomers]float64

type counter struct {
	sync.Mutex
	value int
}

func balk(i int) {
	//fmt.Println("Customer", i, "balks.")
}

func getHairCut(i int) {
	//fmt.Println("Customer", i, "gets a haircut.")
	//time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond) // Simulates time to get haircut.
}

func cutHair() {
	//time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond) // Simulates time to give a haircut.
}

func barberThread(customer chan int, barber chan int, customerDone chan int, barberDone chan int) {
	for {
		customerNum := <-customer
		barber <- customerNum
		cutHair()
		customerNum = <-customerDone
		barberDone <- customerNum
	}
}

func customerThread(i int, numCustomers *counter, customer chan int, barber chan int, customerDone chan int, barberDone chan int, wg *sync.WaitGroup) {
	//fmt.Println("Customer", i, "enters the barbershop.")
	start := time.Now()
	numCustomers.Lock()
	if numCustomers.value == maxCustomers {
		numCustomers.Unlock()
		balk(i)
		wg.Done()
		return
	}
	numCustomers.value++
	numCustomers.Unlock()

	customer <- i
	<-barber
	end := time.Now()
	getHairCut(i)
	customerDone <- i
	<-barberDone

	numCustomers.Lock()
	numCustomers.value--
	//fmt.Println("Customer", i, "leaves the barbershop.")
	numCustomers.Unlock()
	waitingTimes[i] = end.Sub(start).Seconds()
	wg.Done()
}

func main() {
	var durations float64
	var averageWaitingTimes float64
	for j := 0; j < numTrials; j++ {
		var wg sync.WaitGroup
		var numCustomers = &counter{value: 0} // Keeps track of how many customers are in the barbershop.
		customer := make(chan int, 1)
		barber := make(chan int, 1)
		customerDone := make(chan int, 1)
		barberDone := make(chan int, 1)

		start := time.Now()
		go barberThread(customer, barber, customerDone, barberDone)
		wg.Add(totalCustomers)
		for i := 0; i < totalCustomers; i++ {
			//time.Sleep(time.Duration(rand.Intn(100)) * time.Millisecond) // Simulates time between customers arriving.
			go customerThread(i, numCustomers, customer, barber, customerDone, barberDone, &wg)
		}
		wg.Wait()
		end := time.Now()
		durations += end.Sub(start).Seconds()

		// get all non-zero waiting times
		var sum float64
		var len float64
		for i := 0; i < totalCustomers; i++ {
			if waitingTimes[i] != 0 {
				sum += waitingTimes[i]
				len++
				waitingTimes[i] = 0
			}
		}
		if len == 0 {
			averageWaitingTimes = 0
		} else {
			averageWaitingTimes += sum / len
		}
	}
	fmt.Println("Average waiting time:", averageWaitingTimes/numTrials, "s")
	fmt.Println("Average duration:", durations/numTrials, "s")
}

File: barbershop/bs.py
# This implementation uses a synchronized queue to simulate the waiting room.
import threading
import time
import random
import queue

MAX_CUSTOMERS = 4   # Maximum number of customers that can be in the barbershop at the same time.
TOTAL_CUSTOMERS = 100  # Total number of customers that will arrive throughout the program.
NUM_TRIALS = 100

waitTimes = [0]*TOTAL_CUSTOMERS

def balk(i):
    #print("Customer", i, "balks.")
    return

def getHairCut(i):
    #print("Customer", i, "gets a haircut.")
    #time.sleep(0.001*random.randint(0,100)) # Simulate time to get a haircut.
    return

def cutHair():
    #time.sleep(0.001*random.randint(0,100)) # Simulate time to give a haircut.
    return

def barberThread(customer, barber, customerDone, barberDone):
    while True:
        customer.acquire()
        barber.release()
        cutHair()
        customerDone.acquire()
        barberDone.release()

def customerThread(i, waitingRoom, barberChair, customer, barber, customerDone, barberDone):
    #print("Customer", i, "enters the barbershop.")
    global waitTimes
    start = time.time()
    if not barberChair.acquire(False):
        try:
            waitingRoom.put(i, block=False)
            barberChair.acquire()
            waitingRoom.get()
        except queue.Full:
            balk(i)
            return

    customer.release()
    barber.acquire()
    end = time.time()
    getHairCut(i)
    customerDone.release()
    barberDone.acquire()
    barberChair.release()
    #print("Customer", i, "leaves the barbershop.")
    waitTimes[i] = end-start

def main():
    durations = [0]*NUM_TRIALS
    averageWaitTimes = [0]*NUM_TRIALS
    for j in range(NUM_TRIALS):
        waitingRoom = queue.Queue(MAX_CUSTOMERS-1)
        barberChair = threading.Lock()
        customer = threading.Semaphore(0)
        barber = threading.Semaphore(0)
        customerDone = threading.Semaphore(0)
        barberDone = threading.Semaphore(0)

        start = time.time()
        bThread = threading.Thread(target=barberThread, args=(customer, barber, customerDone, barberDone))
        # Make the barber a background thread so the main thread can exit even if it is still running.
        bThread.daemon = True
        bThread.start()

        customers = []
        for i in range(TOTAL_CUSTOMERS):
            #time.sleep(0.001*random.randint(0,100)) # Simulate random amount of time between customers arriving.
            cThread = threading.Thread(target=customerThread, args=(i, waitingRoom, barberChair, customer, barber, customerDone, barberDone))
            customers.append(cThread)
            cThread.start()
        
        for c in customers:
            c.join()

        end = time.time()
        durations[j] = end-start
        global waitTimes
        waitTimes = [value for value in waitTimes if value != 0]
        averageWaitTimes[j] = sum(waitTimes)/len(waitTimes)
        waitTimes = [0]*TOTAL_CUSTOMERS

    print("Average wait time:", sum(averageWaitTimes)/len(averageWaitTimes), "s")
    print("Average duration:", sum(durations)/len(durations), "s")

if __name__ == "__main__":
    main()